File: dataset/Vanilla/GEMINI_gemini-2.5-pro/Task135.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <random>
#include <algorithm>

// Simulates server-side session storage (SessionID -> CSRF_Token)
std::unordered_map<std::string, std::string> session_storage;

/**
 * @brief Generates a random CSRF token.
 *
 * NOTE: For production, use a cryptographically secure random number generator,
 * e.g., from a library like OpenSSL. This is a simplified example.
 *
 * @param length The desired length of the token.
 * @return A random string token.
 */
std::string generateCsrfToken(size_t length = 32) {
    static const std::string chars = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    std::string token;
    token.reserve(length);
    
    std::random_device rd;
    std::mt19937 generator(rd());
    std::uniform_int_distribution<int> distribution(0, chars.length() - 1);

    for (size_t i = 0; i < length; ++i) {
        token += chars[distribution(generator)];
    }
    return token;
}

/**
 * @brief Stores the CSRF token in the simulated session.
 * @param sessionId The user's session identifier.
 * @param token The CSRF token to store.
 */
void storeTokenInSession(const std::string& sessionId, const std::string& token) {
    session_storage[sessionId] = token;
}

/**
 * @brief Validates the received token against the one stored in the session.
 * 
 * NOTE: This uses a simple string comparison. For enhanced security against
 * timing attacks, a constant-time comparison function should be used.
 * 
 * @param sessionId The user's session identifier.
 * @param receivedToken The token received from the client.
 * @return true if the tokens match, false otherwise.
 */
bool validateCsrfToken(const std::string& sessionId, const std::string& receivedToken) {
    if (receivedToken.empty() || sessionId.empty()) {
        return false;
    }
    
    auto it = session_storage.find(sessionId);
    if (it == session_storage.end()) {
        return false; // No token stored for this session
    }
    
    const std::string& storedToken = it->second;
    return storedToken == receivedToken;
}

/**
 * @brief Simulates processing a protected action.
 * @param sessionId The user's session identifier.
 * @param receivedToken The token received from the client.
 */
void processRequest(const std::string& sessionId, const std::string& receivedToken) {
    std::cout << "Processing request for session: " << sessionId << " with token: " << receivedToken << std::endl;
    if (validateCsrfToken(sessionId, receivedToken)) {
        std::cout << "SUCCESS: CSRF token is valid. Action allowed." << std::endl;
    } else {
        std::cout << "FAILURE: CSRF token is invalid or missing. Action rejected." << std::endl;
    }
    std::cout << "-------------------------------------------------" << std::endl;
}

int main() {
    // --- Test Case 1: Valid Request ---
    std::cout << "Test Case 1: Valid Request" << std::endl;
    std::string sessionId1 = "user_session_abc123";
    std::string token1 = generateCsrfToken();
    storeTokenInSession(sessionId1, token1);
    // User submits the form with the correct token
    processRequest(sessionId1, token1);

    // --- Test Case 2: Invalid/Incorrect Token ---
    std::cout << "Test Case 2: Invalid/Incorrect Token" << std::endl;
    std::string sessionId2 = "user_session_def456";
    std::string token2 = generateCsrfToken();
    storeTokenInSession(sessionId2, token2);
    // Attacker tries to submit the form with a guessed/wrong token
    processRequest(sessionId2, "wrong_token_-_");
    
    // --- Test Case 3: Missing Token ---
    std::cout << "Test Case 3: Missing Token" << std::endl;
    std::string sessionId3 = "user_session_ghi789";
    std::string token3 = generateCsrfToken();
    storeTokenInSession(sessionId3, token3);
    // Attacker's request is missing the token
    processRequest(sessionId3, "");

    // --- Test Case 4: No Token in Session ---
    std::cout << "Test Case 4: No Token in Session" << std::endl;
    std::string sessionId4 = "user_session_jkl012";
    // No token is generated or stored for this session
    // A request comes in with some token
    processRequest(sessionId4, "some_random_token");

    // --- Test Case 5: Token Mismatch between Users ---
    std::cout << "Test Case 5: Token Mismatch between Users" << std::endl;
    std::string userA_sessionId = "user_A_session";
    std::string userA_token = generateCsrfToken();
    storeTokenInSession(userA_sessionId, userA_token);
    
    std::string userB_sessionId = "user_B_session";
    std::string userB_token = generateCsrfToken();
    storeTokenInSession(userB_sessionId, userB_token);

    std::cout << "Attacker tries to use User B's token for User A's session:" << std::endl;
    // Attacker on a different session tries to use User B's token to attack User A
    processRequest(userA_sessionId, userB_token);

    return 0;
}

File: dataset/Vanilla/GEMINI_gemini-2.5-pro/Task135.py
import secrets
import os

# Simulates server-side session storage (session_id -> csrf_token)
_session_storage = {}

def generate_csrf_token():
    """Generates a secure, random token."""
    return secrets.token_hex(16)

def store_token_in_session(session_id, token):
    """Stores the CSRF token in the simulated session."""
    _session_storage[session_id] = token

def validate_csrf_token(session_id, received_token):
    """
    Validates the received token against the one stored in the session.
    Uses secrets.compare_digest for constant-time comparison to mitigate timing attacks.
    """
    if not received_token or not session_id:
        return False
    
    stored_token = _session_storage.get(session_id)
    if not stored_token:
        return False
        
    return secrets.compare_digest(stored_token, received_token)

def process_request(session_id, received_token):
    """Simulates processing a protected action."""
    print(f"Processing request for session: {session_id} with token: {received_token}")
    if validate_csrf_token(session_id, received_token):
        print("SUCCESS: CSRF token is valid. Action allowed.")
    else:
        print("FAILURE: CSRF token is invalid or missing. Action rejected.")
    print("-------------------------------------------------")

def main():
    # --- Test Case 1: Valid Request ---
    print("Test Case 1: Valid Request")
    session_id1 = "user_session_abc123"
    token1 = generate_csrf_token()
    store_token_in_session(session_id1, token1)
    # User submits the form with the correct token
    process_request(session_id1, token1)

    # --- Test Case 2: Invalid/Incorrect Token ---
    print("Test Case 2: Invalid/Incorrect Token")
    session_id2 = "user_session_def456"
    token2 = generate_csrf_token()
    store_token_in_session(session_id2, token2)
    # Attacker tries to submit the form with a guessed/wrong token
    process_request(session_id2, "wrong_token_-_")
    
    # --- Test Case 3: Missing Token ---
    print("Test Case 3: Missing Token")
    session_id3 = "user_session_ghi789"
    token3 = generate_csrf_token()
    store_token_in_session(session_id3, token3)
    # Attacker's request is missing the token
    process_request(session_id3, None)

    # --- Test Case 4: No Token in Session ---
    print("Test Case 4: No Token in Session")
    session_id4 = "user_session_jkl012"
    # No token is generated or stored for this session
    # A request comes in with some token
    process_request(session_id4, "some_random_token")

    # --- Test Case 5: Token Mismatch between Users ---
    print("Test Case 5: Token Mismatch between Users")
    userA_session_id = "user_A_session"
    userA_token = generate_csrf_token()
    store_token_in_session(userA_session_id, userA_token)
    
    userB_session_id = "user_B_session"
    userB_token = generate_csrf_token()
    store_token_in_session(userB_session_id, userB_token)

    print("Attacker tries to use User B's token for User A's session:")
    # Attacker on a different session tries to use User B's token to attack User A
    process_request(userA_session_id, userB_token)

if __name__ == "__main__":
    main()